### PR TITLE
Optional report uniqueing in the filters

### LINKED ISSUE #811: Optional report uniqueing in the filters
The extended filtering in #687 should have an optional report uniqueing tick box.

Reports that have the same bug hash identifier should be counted once only in the report list when this option is switched on.
With this feature it will be possible to count the same reports occurring in different runs only once.

Counting and querying the results should support this.
New UI button or checkbox should be available for the user to turn this option on/off.

In this unique listing mode return distinct results and the following columns only:
bughash, filename, chekcer_id, severity, review status
`select distinct checker_id,checker_cat, bug_id,review_statuses.status from reports join files on files.id=file_id left outer join review_statuses on bug_id=review_statuses.bug_hash;`
Then from the report listing we can create links to the bug viewer based on bug hash.
In the permalinks we could place the bug hash and show all bugs with the same hash (similar bugs) in the bug viewer. This feature depends on #620 

All existing filters can remain the same (because we can also use the columns in the where clause which were not selected by distinct).

### PR DESCRIPTION
**Pull request by csordasmarton**
Closes #811

### CODE REVIEW

File: libcodechecker/server/client_db_access_handler.py [hunk 1]
```diff
@@ -758,12 +787,24 @@ def getRunResultCount(self, run_ids, report_filter, cmp_data):
...
-                           ReviewStatus.bug_hash == Report.bug_id) \
-                .filter(filter_expression)
+            isUnique = report_filter is not None and \
+                       report_filter.isUnique
+
+            if isUnique:
+                q = session.query(Report.bug_id,
+                                  Report.checker_id,
+                                  Report.severity) \
+                    .distinct(Report.bug_id,
```
Comment (gyorb) [inline libcodechecker/server/client_db_access_handler.py]: Shouldn't we use the same uniqueing here as before?
` .distinct(Report.bug_id, Report.checker_id, Report.checker_message, Report.severity, File.filename, ReviewStatus.status)`

File: www/scripts/codecheckerviewer/BugFilterView.js [hunk 2]
```diff
@@ -1407,6 +1448,28 @@ function (declare, Deferred, dom, domClass, all, topic, Standby, Button,
     },
 
     /**
+     * Deselecting items which are not active by the URL state.
```
Comment (gyorb) [inline www/scripts/codecheckerviewer/BugFilterView.js]: Do we need this in this pull request? It does nothing right now.

File: libcodechecker/server/client_db_access_handler.py (hunk 1 above)
Comment (csordasmarton) [reply libcodechecker/server/client_db_access_handler.py]: Yes, we have to. Thanks.

File: www/scripts/codecheckerviewer/BugFilterView.js (hunk 2 above)
Comment (csordasmarton) [reply www/scripts/codecheckerviewer/BugFilterView.js]: Yes, we need this. I didn't have time to write this function last night, I wrote it this morning and I removed the WIP label from this pull request.

File: api/v6/report_server.thrift [hunk 3]
```diff
@@ -200,10 +201,12 @@ typedef list<CommentData> CommentDataList
 
 /**
  * Members of this struct are interpreted in "AND" relation with each other.
+ * Between the members there is "OR" relation.
```
Comment (Xazax-hun) [inline api/v6/report_server.thrift]: I think this comment is confusing. Maybe reword to something like between the elements of the list.

File: api/v6/report_server.thrift [hunk 4]
```diff
@@ -200,10 +201,12 @@ typedef list<CommentData> CommentDataList
...
  */
 struct RunFilter {
-  1: list<i64> runIds,
-  2: string    name
+  1: list<i64>    ids,        // Run ids of the runs.
```
Comment (Xazax-hun) [inline api/v6/report_server.thrift]: Redundant info in this comment, reword like `IDs of the runs.`

File: api/v6/report_server.thrift [hunk 5]
```diff
@@ -427,4 +430,4 @@ service codeCheckerDBAccess {
                    5: bool   force)
                    throws (1: shared.RequestFailed requestError),
 
-}
+}
```
Comment (Xazax-hun) [inline api/v6/report_server.thrift]: I think this change is irrelevant.

File: libcodechecker/server/client_db_access_handler.py [hunk 6]
```diff
@@ -611,7 +619,7 @@ def getReport(self, reportId):
 
     @timeit
     def getRunResults(self, run_ids, limit, offset, sort_types,
-                      report_filter, cmp_data):
+                         report_filter, cmp_data):
```
Comment (Xazax-hun) [inline libcodechecker/server/client_db_access_handler.py]: The formatting seems to be off here.

File: libcodechecker/server/client_db_access_handler.py [hunk 7]
```diff
@@ -640,51 +648,72 @@ def getRunResults(self, run_ids, limit, offset, sort_types,
...
+                       report_filter.isUnique
+
+            if isUnique:
+                q = session.query(Report.bug_id,
+                                  Report.checker_id,
+                                  Report.checker_message,
+                                  Report.severity,
+                                  File.filename,
+                                  ReviewStatus) \
+                           .distinct(Report.bug_id,
```
Comment (Xazax-hun) [inline libcodechecker/server/client_db_access_handler.py]: I would leave a TODO here. Is this deterministic which bug will be chosen from the ones with the same bug id? Maybe in the future, we would like to get the one with the shortest path or have some other criterion.

File: libcodechecker/server/client_db_access_handler.py [hunk 8]
```diff
@@ -758,12 +787,30 @@ def getRunResultCount(self, run_ids, report_filter, cmp_data):
                     # There is no difference.
                     return 0
 
-            q = session.query(Report) \
-                .filter(Report.run_id.in_(run_ids)) \
-                .outerjoin(File, Report.file_id == File.id) \
-                .outerjoin(ReviewStatus,
-                           ReviewStatus.bug_hash == Report.bug_id) \
-                .filter(filter_expression)
+            isUnique = report_filter is not None and \
```
Comment (Xazax-hun) [inline libcodechecker/server/client_db_access_handler.py]: This code is very similar to the one in `getRunResults`. Maybe it makes sense to hoist some of the functionality into functions to reduce the duplication?

File: libcodechecker/server/run_db_model.py [hunk 9]
```diff
@@ -103,7 +104,8 @@ class File(Base):
     __tablename__ = 'files'
 
     id = Column(Integer, autoincrement=True, primary_key=True)
-    filepath = Column(String)
+    filepath = Column(String, nullable=False)
+    filename = Column(String, nullable=False)
```
Comment (Xazax-hun) [inline libcodechecker/server/run_db_model.py]: Does it make sense to store the filename separately? Don't we have some functionality to compute this on demand (on the SQL server)? E.g.: see https://stackoverflow.com/questions/19505875/parse-file-name-and-path-from-full-path
